File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean solve(TreeNode root, long min, long max){
        if(root == null)
            return true;
        
        if((long)root.val > min && (long)root.val < max){
            boolean left = solve(root.left, min, (long)root.val);
            boolean right = solve(root.right, (long)root.val, max);
            return left && right;
        }
        else 
            return false;
    }
    public boolean isValidBST(TreeNode root) {
        return solve(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool solve(TreeNode* root, long leftMax, long rightMax){
        if(!root)
            return true;
        if(root->val >= rightMax || root->val <= leftMax)
            return false;
        return solve(root->left, leftMax, root->val) && solve(root->right, root->val, rightMax);
    }
    bool isValidBST(TreeNode* root) {
        long leftMax = LONG_MIN;
        long rightMax = LONG_MAX;
        return solve(root, leftMax, rightMax);
    }
};